error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MemberOrdering.java

```diff
@@ -0,0 +1,187 @@
+        .build();
+  }
+
+  private static boolean isStatic(VariableTree memberTree) {
+    Set<Modifier> modifiers = memberTree.getModifiers().getFlags();
+    return modifiers.contains(Modifier.STATIC);
+  }
+
+  private static boolean isConstructor(MethodTree methodDecl) {
+    return ASTHelpers.getSymbol(methodDecl).isConstructor();
+  }
+
+  private static boolean shouldBeSorted(Tree tree) {
+    return tree instanceof VariableTree
+        || (tree instanceof MethodTree && !ASTHelpers.isGeneratedConstructor((MethodTree) tree));
+  }
+
+  private static SuggestedFix swapMembersWithComments(
+      ImmutableList<ClassMemberWithComments> memberWithComments,
+      ImmutableList<ClassMemberWithComments> sortedMembersWithComments,
+      VisitorState state) {
+    SuggestedFix.Builder fix = SuggestedFix.builder();
+    for (int i = 0; i < memberWithComments.size(); i++) {
+      Tree originalMember = memberWithComments.get(i).member();
+      ClassMemberWithComments correct = sortedMembersWithComments.get(i);
+      /* Technically this check is not necessary, but it avoids redundant replacements. */
+      if (!originalMember.equals(correct.member())) {
+        String replacement =
+            Stream.concat(
+                    correct.comments().stream().map(Tokens.Comment::getText),
+                    Stream.of(state.getSourceForNode(correct.member())))
+                .collect(joining("\n"));
+        fix.merge(
+            SuggestedFixes.replaceIncludingComments(
+                TreePath.getPath(state.getPath(), originalMember), replacement, state));
+      }
+    }
+    return fix.build();
+  }
+
+  /** Returns the class' members with their comments. */
+  private static ImmutableList<ClassMemberWithComments> getMembersWithComments(
+      ClassTree classTree, VisitorState state) {
+    return classTree.getMembers().stream()
+        .map(
+            member ->
+                new ClassMemberWithComments(member, getMemberComments(state, classTree, member)))
+        .collect(toImmutableList());
+  }
+
+  private static ImmutableList<Tokens.Comment> getMemberComments(
+      VisitorState state, ClassTree classTree, Tree member) {
+    if (member.getKind() == Tree.Kind.METHOD
+        && ASTHelpers.isGeneratedConstructor((MethodTree) member)) {
+      return ImmutableList.of();
+    }
+
+    checkState(
+        state.getEndPosition(member) != -1,
+        "Member's end position is not available (-1).\n - member=[%s]\n - source=[%s]",
+        member,
+        state.getSourceForNode(member));
+
+    ImmutableList<ErrorProneToken> tokens =
+        ImmutableList.copyOf(
+            state.getOffsetTokens(
+                ASTHelpers.getStartPosition(classTree), state.getEndPosition(classTree)));
+
+    int memberStartPos = ASTHelpers.getStartPosition(member);
+    Optional<Integer> previousMemberEndPos =
+        tokens.stream()
+            .map(ErrorProneToken::endPos)
+            .takeWhile(endPos -> endPos <= memberStartPos)
```

Comment:
I did some investigation on the `<` or `<=` question. I don't think, it matters too much which way we go. It would matter if we could place a comment on a token that immediately follows another. This is impossible because if we comment a token, the token immediately follows its comment instead of its prior token.

```
// `int b` immidately follows the token `;`
int a;int b;
// `int b` doesn't follow immidately any error-prone token.
int a;/* comment */int b;
```
An interesting case would be a code like this:
```
...
/* comment */
;void method() { ... }
```
In this case, the comment belongs to the token `;` and not the method. Our current implementation will move the comment with the `void method() { ... }` instead of leaving it with the `;` token. This aligns with the implementation of `SuggestedFixes#replaceIncludingComments` and it provides a better UX than the alternative, but AFAIK it is technically incorrect.

I've added a test case for this, but again, this will be resolved by the simplest formatters.

Reply:
Interesting case! 
I originally used `<=` over `<`, as I figured there was no reason to exclude the start position of the token, if for some reason a previous token has that as end position.